pkg/controller/seed-controller-manager/mla/project_controller.go

```diff
@@ -0,0 +1,228 @@
+	org, err := r.grafanaClient.GetOrgByOrgName(ctx, getOrgNameForProject(project))
+	if err != nil {
+		return err
+	}
+	_, err = r.grafanaClient.DeleteOrg(ctx, org.ID)
+	if err != nil {
+		return err
+	}
+	kubernetes.RemoveFinalizer(project, mlaFinalizer)
+	if err := r.Update(ctx, project); err != nil {
+		return fmt.Errorf("updating Project: %w", err)
+	}
+	return nil
+}
+
+func (r *projectReconciler) createGrafanaOrg(ctx context.Context, org grafanasdk.Org) (grafanasdk.Org, error) {
+	status, err := r.grafanaClient.CreateOrg(ctx, org)
+	if err != nil {
+		return org, fmt.Errorf("unable to add organization: %w (status: %s, message: %s)",
+			err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+	}
+	if status.OrgID == nil {
+		// possibly organization already exists
+		org, err := r.grafanaClient.GetOrgByOrgName(ctx, org.Name)
+		if err != nil {
+			return org, fmt.Errorf("unable to get organization by name %s", org.Name)
+		}
+		return org, nil
+	}
+	org.ID = *status.OrgID
+	return org, nil
+}
+
+func (r *projectReconciler) ensureOrganization(ctx context.Context, log *zap.SugaredLogger, project *kubermaticv1.Project, expected grafanasdk.Org, annotationKey string) error {
+	orgID, ok := project.GetAnnotations()[annotationKey]
+	if !ok {
+		org, err := r.createGrafanaOrg(ctx, expected)
+		if err != nil {
+			return fmt.Errorf("unable to create grafana org: %w", err)
+		}
+		if err := r.setAnnotation(ctx, project, annotationKey, strconv.FormatUint(uint64(org.ID), 10)); err != nil {
+			// revert org creation, if deletion failed, we can't do much about it
+			// if we failed at this moment and the project would be renamed quickly, that organization will be orphaned and we will never remove it.
+			if status, err := r.grafanaClient.DeleteOrg(ctx, org.ID); err != nil {
+				log.Debugf("unable to delete organization: %w (status: %s, message: %s)",
+					err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+			}
+			return err
+		}
+		return nil
+	}
+	id, err := strconv.ParseUint(orgID, 10, 32)
+	if err != nil {
+		return err
+	}
+
+	org, err := r.grafanaClient.GetOrgById(ctx, uint(id))
+	if err != nil {
+		// possibly not found
+		org, err := r.createGrafanaOrg(ctx, expected)
+		if err != nil {
+			return fmt.Errorf("unable to create grafana org: %w", err)
+		}
+		if err := r.setAnnotation(ctx, project, annotationKey, strconv.FormatUint(uint64(org.ID), 10)); err != nil {
+			// revert org creation, if deletion failed, we can't do much about it
+			if status, err := r.grafanaClient.DeleteOrg(ctx, org.ID); err != nil {
+				log.Debugf("unable to delete organization: %w (status: %s, message: %s)",
+					err, pointer.StringPtrDerefOr(status.Status, "no status"), pointer.StringPtrDerefOr(status.Message, "no message"))
+			}
+			return err
+		}
+		return nil
+	}
+	expected.ID = uint(id)
+	if !reflect.DeepEqual(org, expected) {
```

Comment:
I have some concerns regarding `DeepEqual`. The `Org` struct has three fields: ID, Name, and Address.

`expected` has ID and Name set. We expect that `org` will have ID and Name set, but what about the Address? Since `org` is coming from the API, isn't it going to have Address as well? If it has the Address, isn't it going to fail DeppEqual? In that case, we would do an update on each reconciliation loop, which can be avoided.

Reply:
only we are creating those orgs and we don't set address, so I think we should be fine here